### ISSUE TITLE
ValueError: _jpype.__spec__ is None when running main.exe

### ISSUE THREAD

**Issue by Sooraj-beep** [Nuitka/Nuitka#1405]
Before submitting an Issue, please review the [Issue Guidelines](https://github.com/kayhayen/Nuitka/blob/develop/CONTRIBUTING.md#submitting-an-issue).

* Nuitka version:
```
name:  None
fullname:  nuitka
name:  None
fullname:  nuitka.__main__
0.6.20rc4
Commercial: None
Python: 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:19:05) [MSC v.1916 64 bit (AMD64)]
Flavor: Anaconda Python
Executable: C:\Users\sooraj.a\Anaconda3\envs\pyimagej\python.exe
OS: Windows
Arch: x86_64
```
* How did you install Nuitka and Python: I used pip inside an anaconda environment.
* To setup my environment and include all modules I use: 
```
conda create -n pyimagej -c conda-forge openjdk=8 pyimagej
conda activate pyimagej

```
* The specific PyPI names and versions

  It should be taken from this output if there specific packages involved, e.g.
  `numpy`, you are expected to shorten this to the relevant ones.

  > python -m pip freeze
name:  None
fullname:  pip
name:  None
fullname:  pip.__main__
certifi==2021.10.8
cycler @ file:///home/conda/feedstock_root/build_artifacts/cycler_1635519461629/work
fonttools @ file:///D:/bld/fonttools_1643090960075/work
imglyb @ file:///home/conda/feedstock_root/build_artifacts/imglyb_1632508347959/work
importlib-metadata @ file:///D:/bld/importlib-metadata_1642514138647/work
jgo @ file:///home/conda/feedstock_root/build_artifacts/jgo_1632400948459/work
JPype1 @ file:///D:/bld/jpype1_1636588042576/work
kiwisolver @ file:///D:/bld/kiwisolver_1635836839743/work
matplotlib @ file:///D:/bld/matplotlib-suite_1639359275746/work
munkres==1.1.4
Nuitka @ https://github.com/Nuitka/Nuitka/archive/develop.zip
numpy @ file:///D:/bld/numpy_1643009390959/work
olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1637239678211/work
pandas @ file:///D:/bld/pandas_1642871738130/work
Pillow @ file:///D:/bld/pillow_1636558975346/work
psutil @ file:///D:/bld/psutil_1640887336249/work
pyimagej @ file:///D:/bld/pyimagej_1636821600677/work
pyparsing @ file:///home/conda/feedstock_root/build_artifacts/pyparsing_1642753572664/work
PyQtWebEngine==5.15.4
PyQtWebEngine-Qt5==5.15.2
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1626286286081/work
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1633452062248/work
scyjava @ file:///home/conda/feedstock_root/build_artifacts/scyjava_1638980239838/work
six @ file:///home/conda/feedstock_root/build_artifacts/six_1620240208055/work
typing_extensions @ file:///home/conda/feedstock_root/build_artifacts/typing_extensions_1638334978229/work
unicodedata2 @ file:///D:/bld/unicodedata2_1640031186081/work
xarray @ file:///home/conda/feedstock_root/build_artifacts/xarray_1643438115178/work
zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1641731510635/work

* My code is in main.py which looks like this:
```
import imagej
import os
ij = imagej.init('2.1.0')
print(ij.getVersion())
ij.dispose()
```
* then I ran the following nuitka command:
`nuitka --standalone --show-progress --enable-plugin=numpy --include-data-file=C:\Users\sooraj.a\Anaconda3\envs\pyimagej\Lib\site-packages\org.jpype.jar=.\org.jpype.jar  .\main.py`

* now I run main.exe inside main.dist and get the following error:
```
Traceback (most recent call last):
  File "C:\Users\sooraj.a\WORK_S~1\TESTNU~1\MAIN~2.DIS\main.py", line 9, in <module>
  File "C:\Users\sooraj.a\WORK_S~1\TESTNU~1\MAIN~2.DIS\imagej\__init__.py", line 165, in init
  File "C:\Users\sooraj.a\WORK_S~1\TESTNU~1\MAIN~2.DIS\scyjava\__init__.py", line 59, in start_jvm
  File "C:\Users\sooraj.a\WORK_S~1\TESTNU~1\MAIN~2.DIS\jpype\_core.py", line 226, in startJVM
  File "C:\Users\sooraj.a\WORK_S~1\TESTNU~1\MAIN~2.DIS\importlib\util.py", line 120, in find_spec
ValueError: _jpype.__spec__ is None
```

**Comment by kayhayen** [comment]
Can you make sure that your example if not referring to things I cannot have? These absolute paths appear to contain things I would have to have at least a clear instruction set on how to install them. The crash appears to be in `imagej.init` and use a file path that I wouldn't have.

**Comment by Sooraj-beep** [comment]
Thanks for replying. I have updated my issue to include information on how to setup my environment. Do let me know if any other information is required.

**Comment by kayhayen** [comment]
Are you saying, that the `conda install` is going to give me a folder named `C:\InterfaceFluidics\AADependencies` ?

**Comment by Sooraj-beep** [comment]
and to answer your question, conda install does not give that folder. just replace that path with path to your JDK.

**Comment by Sooraj-beep** [comment]
I also noticed that the nuitka command has to be changed to point to the jar file in your system as well. Hopefully my path will give you an idea to find where in your Anaconda folder you can find the required jar file.

**Comment by Sooraj-beep** [comment]
I have updated my issue to not include things you may not have.

**Comment by Sooraj-beep** [comment]
You will still have to include the org.jpype.jar file in your Nuitka command. Hopefully the path in my issue will help you in locating the jar file in your system. Thanks

**Comment by kayhayen** [comment]
It seems to hang after startup, potentially downloading stuff, to where I don't know. I assume this is not running it as a server, but supposed to print the version and then quit, right?

**Comment by kayhayen** [comment]
Ok, so, I guess, you are saying, it downloads to the site-packages folder, a jar file, and when that's present, it will just use it, and that is why you included that as a data file, seems reasonable. I would like to criticize "them" for not putting it below their package, but top level, and this may well be where the problem lives.

**Comment by Sooraj-beep** [comment]
> It seems to hang after startup, potentially downloading stuff, to where I don't know. I assume this is not running it as a server, but supposed to print the version and then quit, right?

Yes, this is what it is supposed to do. Unsure if its downloading anything or just running on server (I always assumed its the latter).

**Comment by Sooraj-beep** [comment]
> Ok, so, I guess, you are saying, it downloads to the site-packages folder, a jar file, and when that's present, it will just use it, and that is why you included that as a data file, seems reasonable. I would like to criticize "them" for not putting it below their package, but top level, and this may well be where the problem lives.

Yes, exactly. This was my conclusion too but I lack the knowledge to figure out how to fix this issue regarding __spec__ files.

**Comment by kayhayen** [comment]
So am I reproducing this indeed with Anaconda on Windows. I am a bit surprised it uses pandas, as that is making the compilation a lot larger than I would hope for. I am therefore trying to remove that as a dependency, it's not listed in their code I think, although with 666 modules it's not as bad as it used to be anymore. It might use multiprocessing, I am not sure of that, but it would make sense.

It appears to me, that this could be about a too late assignment of `__spec__` for modules that load in two phases, as newer Python allows for extension modules. These appear to become more common, and some expect that value, like `__file__` in a recent issue, are already correct when starting their execution. Should this prove correct, I am hoping to include it in the next hotfix of Nuitka even.

**Comment by kayhayen** [comment]
So, I monkey patched the `find_spec` like this in `jpype._core`:

```
o_find_spec = _util.find_spec

def my_find_spec(*args, **kwargs):
    print("FIND_SPEC", args, kwargs)
    r = o_find_spec(*args, **kwargs)
    print("RES", r)
    return r

_util.find_spec = my_find_spec
```

This was to find out what the C code is doing there. The search is for `_jpype` which is indeed an extension module, but it's the one searched.

**Comment by kayhayen** [comment]
And adding a traceback, this is indeed, `_jpype` asking for the spec of *itself*, and it's not doing that by looking up the `__spec__` value, which arguably could be considered semi-correct. Question now is, if our meta path based loader is asked at all, since of course it has a `find_spec` in the loader. Very probably it's not used, because the `__loader__` attribute would be set late.

**Comment by kayhayen** [comment]
Nah, confusing myself there, this is the code:

```
        module = sys.modules[fullname]
        if module is None:
            return None
        try:
            spec = module.__spec__
        except AttributeError:
            raise ValueError('{}.__spec__ is not set'.format(name)) from None
        else:
            if spec is None:
                raise ValueError('{}.__spec__ is None'.format(name))
            return spec
```

Interestingly, we *never* so far set the `__spec__` value for extension modules. This is independent of one or two phase loading method. Seems like yet another incompatibility was found. The `__spec__` value handling needs to be added in much the same way as we do it for compiled modules already.

**Comment by kayhayen** [comment]
Interestingly we create the module in 2 phase load through a spec value with `PyModule_FromDefAndSpec` which one could expect to set the value of `__spec__`, but so far I didn't find in the CPython code where it would, so probably it does not and it would explain what we see. Added a one liner that sets the __spec__ attribute as created. I think this might make the issue go away, but could need more handling of the `initializing` attribute, to handle concurrent imports.

**Comment by kayhayen** [comment]
So, I added setting the `__spec__` for the two phase loading path, where clearly it was missing, but I didn't get an improvement out of that. I am currently verifying if that is what it is doing anyway. Adding more traces, it's a bit unclear when exactly the C code is doing its thing. 

But I noticed in the mean time that one phase loading fallback of new style importing, also has an issue, that already shows with this kind of code:

```
import _ctypes
print(_ctypes.__spec__)
```

This is using the fallback, even with the newest Python. And the spec is `None` in this case too. So we need to look into fixing that one, of course, too. If I am not mistaken, that's not our code path, but obviously incompatibilities like that must not remain, and maybe, just maybe I was mistaken and it's actually one phase too for `_jpype`.

**Comment by kayhayen** [comment]
Also, setting the `origin` value in the spec value, just in case. The `__file__` was already correct for one phase loading fallback, but some things might use that. Still not sure what `_jpype` actually wants to do with that `__spec__` value once it has it.

**Comment by kayhayen** [comment]
It seems, that `_jype` is using 2 phase loading, and then doing stuff you are not supposed to do in phase 1, so Nuitka has to set it a bit earlier.

**Comment by kayhayen** [comment]
When retrying it with current Anaconda, I was getting a hanging process, executing some subprocess apparently, so I cannot really be sure with what it is doing. I found an issue with `scyjava` and metadata, and will be compiling until I see it hang equally well. The 2 phase loading issues should have been fixed a while ago already in other issues. So actually there is hope. But I guess it may be that I remain unable to confirm. Not sure how to debug that hang from your example, or is that expected. I have used it before, or tried at least, and I think I would have noted if it hung, but I may well have failed to execute it. Using it from PyPI seems hopeless.

**Comment by kayhayen** [comment]
I managed to get it to compile and run, adding a few data files and dynamic metadata version workarounds for used packages, with that ìmagej` and `pjype` were working.

**Comment by kayhayen** [comment]
Thanks for your report, this is fixed on the factory branch, which is a development version under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 

Feedback if this is working is very welcome, just please do not share plans of doing it, but rather confirmations or denials of it working.

**Comment by kayhayen** [comment]
This is now on develop branch and will be part of 1.5 release shortly.

**Comment by kayhayen** [comment]
This is part of the stable release 1.5 that I just made.
